lib/room.rb

```diff
@@ -0,0 +1,101 @@
+require_relative "reservation"
+require_relative "csv_record"
+
+module HotelGroup
+  class Room < CsvRecord
+    attr_accessor :number, :price, :reservations, :unavailable_dates, :blocks, :block_price
+
+    def initialize(number, price)
+      @number = number
+      @price = price
+      @block_price = price * 0.8
+      @unavailable_dates = []
+      @reservations = []
+      @blocks = []
+    end
+
+    # def self.make_rooms_list(number, price)
+    #   rooms = []
+    #   number.times do |n|
+    #     room = Room.new(n + 1, price)
+
+    #     rooms << room
+    #   end
+    #   return rooms
+    # end
+
+    def connect(res)
+      reservations << res
+      set_unavailable(res.start_time, res.end_time)
+      return self
+    end
+
+    def add_reservation(res)
+      if has_reservation?(res.start_time, res.end_time)
+        raise ArgumentError, "Room #{number} is already reserved for this date range"
+      end
+      reservations << res
+      set_unavailable(res.start_time, res.end_time)
+    end
+
+    def is_available?(start_time, end_time)
+      if unavailable_dates == []
+        return true
+      end
+      unavailable_dates.each do |date_array|
+        res_start = date_array[0]
+        res_end = date_array[1]
+
+        if (res_start...res_end).include?(start_time) || (res_start + 1..res_end).include?(end_time) ||
+           (start_time...end_time).include?(res_start) ||
+           (start_time + 1..end_time).include?(res_end)
+          return false
+        end
+      end
+      return true
+    end
+
+    def set_unavailable(start_time, end_time)
+      unavailable_dates << [start_time, end_time]
```

Comment:
This seems like you're starting to have the room know information that seems to be more of interest to the reservation. If a room needs to know when its available, it's probably better for it to have a way to ask its reservations, no?